network/lwip2transport/udp.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lwip2transport
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	lwipLib "github.com/eycorsican/go-tun2socks/core"
+)
+
+type udpHandler struct {
+	// Protects the connections map
+	sync.Mutex
+
+	// Used to establish connections to the proxy
+	listener transport.PacketListener
+
+	// How long to wait for a packet from the proxy. Longer than this and the connection
+	// is closed.
+	timeout time.Duration
+
+	// Maps connections from TUN to connections to the proxy.
+	conns map[lwipLib.UDPConn]net.PacketConn
+}
+
+// newUDPHandler returns a lwIP UDP connection handler.
+//
+// `pl` is an UDP packet listener.
+// `timeout` is the UDP read and write timeout.
+func newUDPHandler(pl transport.PacketListener, timeout time.Duration) *udpHandler {
+	return &udpHandler{
+		listener: pl,
+		timeout:  timeout,
+		conns:    make(map[lwipLib.UDPConn]net.PacketConn, 8),
+	}
+}
+
+func (h *udpHandler) Connect(tunConn lwipLib.UDPConn, target *net.UDPAddr) error {
+	proxyConn, err := h.listener.ListenPacket(context.Background())
+	if err != nil {
+		return err
+	}
+	h.Lock()
+	h.conns[tunConn] = proxyConn
```

Comment:
TODO: Use local_ip:port as the key to map to transport listeners. Creating one listener per 4-tuple is unnecessary.

In that case you will only be able to delete the map entry once all associated proxyConns/relays are closed/done.

Reply:
updated